// powerbi-visuals-utils-dataviewutils/lib/converterHelper.js
// powerbi.extensibility.utils.dataview
import * as DataRoleHelper from "./dataRoleHelper";
export function categoryIsAlsoSeriesRole(dataView, seriesRoleName, categoryRoleName) {
    if (dataView.categories && dataView.categories.length > 0) {
        // Need to pivot data if our category soure is a series role
        const category = dataView.categories[0];
        return category.source &&
            DataRoleHelper.hasRole(category.source, seriesRoleName) &&
            DataRoleHelper.hasRole(category.source, categoryRoleName);
    }
    return false;
}
export function getSeriesName(source) {
    return (source.groupName !== undefined)
        ? source.groupName
        : source.queryName;
}
export function isImageUrlColumn(column) {
    const misc = getMiscellaneousTypeDescriptor(column);
    return misc != null && misc.imageUrl === true;
}
export function isWebUrlColumn(column) {
    const misc = getMiscellaneousTypeDescriptor(column);
    return misc != null && misc.webUrl === true;
}
export function getMiscellaneousTypeDescriptor(column) {
    return column
        && column.type
        && column.type.misc;
}
export function hasImageUrlColumn(dataView) {
    if (!dataView || !dataView.metadata || !dataView.metadata.columns || !dataView.metadata.columns.length) {
        return false;
    }
    return dataView.metadata.columns.some((column) => isImageUrlColumn(column) === true);
}
